tensilelite/Tensile/KernelWriterAssembly.py

```diff
@@ -4125,6 +4125,89 @@ def lraSwapAddressesForDTLPad(self, kernel, tP):
                          comment="xor both lds buffer offsets to enable swapping"))
     return module
 
+  def checkAlignmentForDTLTailLoop(self, kernel, tPA, tPB):
+    # For 16bit/8bit floats we need to check for alignment and modify the
+    # buffer loads when using wider (dword,dwordx4) loads to workaround OOB behaviours
+    #
+    # For 16b:
+    # if k even: nothing needs to be done
+    # else k odd:
+    #   if addr 4B aligned: add 2B to size record in buf desc
+    #   else: shift gro left by 2B and shift lro right by 2B
+    #
+    # For 8b: TBD.
+    #
+    module = Module("Check Alignment For DTL Tail Loop")
+
+    if not kernel["DirectToLdsA"] and not kernel["DirectToLdsB"]:
+      return Module()
+
+    # Skip this for 8bit types for now.
+    if kernel["ProblemType"]["DataTypeA"].isAnyFloat8() or kernel["ProblemType"]["DataTypeB"].isAnyFloat8():
+      return Module()
+
+    module.addComment0("Check if global reads need to be modified for DTL Tail Loop")
+
+    tmpSgpr = self.sgprPool.checkOut(1)
+    labelSkip = Label("SkipDTLTailLoopGRChanges", "" )
+    goodLMultiple = 2
```

Comment:
To completely avoid out of boundary access, goodLMultiple should be GlobalReadVectorWidth.

For example,  GRVW is 4 and K % 4 is 3.
Te sfhit number will be 1 in that case.
However, GlobalRead access 4 continuous elements and it still access out of boundary elements.

Reply:
I think what we essentially need is to avoid out of boundary access.
  mod = K % GRVW
  GlobalReadOffset -= mod * bpeGR
  LocalReadAddr += mod * bpe

I am not sure if we can use wider local read with this logic.

Another thing is that this logic is applicable only for transposed case (TLU=False).
K does not matter for TLU=True case.

Reply:
It might be better to change Srd0 and 2 instead of GlobalReadOffset.
I am not sure if GlobalReadOffset can be negative value or not.

By the way, in case GlobalReadOffset cannot be a negative value, you need to change Srd instead.
However, you cannot just change Srd0.
This is 64bit address.
You should use 64bit calculation instead.